rayon-core/src/scope/mod.rs

```diff
@@ -381,17 +394,70 @@ where
 {
     in_worker(|owner_thread, _| {
         let scope = ScopeFifo::<'scope>::new(owner_thread);
-        unsafe { scope.base.complete(owner_thread, || op(&scope)) }
+        unsafe { scope.base.complete(Some(owner_thread), || op(&scope)) }
     })
 }
 
-impl<'scope> Scope<'scope> {
-    fn new(owner_thread: &WorkerThread) -> Self {
-        Scope {
-            base: ScopeBase::new(owner_thread),
+/// Creates a "fork-join" scope `s` and invokes the closure with a
+/// reference to `s`. This closure can then spawn asynchronous tasks
+/// into `s`. Those tasks may run asynchronously with respect to the
+/// closure; they may themselves spawn additional tasks into `s`. When
+/// the closure returns, it will block until all tasks that have been
+/// spawned into `s` complete.
+///
+/// This is just like `scope()` except the closure runs on the same thread
+/// that calls `external_scope()`. Only work that it spawns runs in the
+/// thread pool.
+///
+/// # Panics
+///
+/// If a panic occurs, either in the closure given to `scope()` or in
+/// any of the spawned jobs, that panic will be propagated and the
+/// call to `scope()` will panic. If multiple panics occurs, it is
+/// non-deterministic which of their panic values will propagate.
+/// Regardless, once a task is spawned using `scope.spawn()`, it will
+/// execute, even if the spawning task should later panic. `scope()`
+/// returns once all spawned jobs have completed, and any panics are
+/// propagated at that point.
+pub fn external_scope<'scope, OP, R>(op: OP) -> R
```

Comment:
I don't love the name 'external scope' -- it may not be external, after all, if this is running on a worker thread, right?

Maybe `in_place_scope` or something like that?

Reply:
`in_place_scope` sounds fine to me but @cuviper should weigh in I guess.

Reply:
`direct_scope`? `immediate_scope`? `local_scope`?

I'm not sure what will be most intuitive, that users might naturally guess what it means compared to `scope` alone.

(My own prior experimentation called it `raw_scope`, but I think that misses the mark.)

Reply:
`this_thread_scope`?

Reply:
`in_place_scope` or `immediate_scope` sound both good. The test is good to explain the use-case.

Reply:
@cuviper can you sign off on `in_place_scope`?

Reply:
Yeah, that's fine.